crates/flux-fhir-analysis/src/wf/mod.rs

```diff
@@ -382,11 +383,53 @@ impl<'a, 'tcx> Wf<'a, 'tcx> {
             .try_collect_exhaust()
     }
 
+    fn check_ty_is_base(&self, ty: &fhir::Ty) -> Result<(), ErrorGuaranteed> {
+        match &ty.kind {
+            fhir::TyKind::BaseTy(_) | fhir::TyKind::Indexed(_, _) => Ok(()),
+            fhir::TyKind::Tuple(tys) => {
+                for ty in tys {
+                    self.check_ty_is_base(ty)?;
+                }
+                Ok(())
+            }
+            fhir::TyKind::Constr(_, ty) | fhir::TyKind::Exists(_, ty) => self.check_ty_is_base(ty),
+
+            fhir::TyKind::Ptr(_, _)
+            | fhir::TyKind::Ref(_, _)
+            | fhir::TyKind::Array(_, _)
+            | fhir::TyKind::RawPtr(_, _)
+            | fhir::TyKind::OpaqueDef(_, _, _, _)
+            | fhir::TyKind::Never
+            | fhir::TyKind::Hole(_) => self.emit_err(errors::InvalidBaseInstance::new(ty)),
+        }
+    }
+
+    fn check_generic_args_kinds(
+        &self,
+        adt_id: Option<DefId>,
+        args: &[fhir::GenericArg],
+    ) -> Result<(), ErrorGuaranteed> {
+        if let Some(def_id) = adt_id &&
+           let Ok(generics) = self.genv.generics_of(def_id)
+        {
+           for (arg, param) in args.iter().zip(generics.params.iter()) {
+             if param.kind == GenericParamDefKind::SplTy {
+                if let fhir::GenericArg::Type(ty) = arg {
+                   self.check_ty_is_base(ty)?;
+                }
+             }
+           }
+        }
+        Ok(())
+    }
+
     fn check_generic_args(
         &mut self,
         infcx: &mut InferCtxt,
+        adt_id: Option<DefId>,
```

Comment:
This function should take `DefId` instead of `Option<DefId>`. For anything with generic arguments, we should be able to get its generic parameters and check the kinds. I'll comment at the respective call site how this function should be called.